integration/setup/bastion_instance.go

```diff
@@ -0,0 +1,559 @@
+// +build k8srequired
+
+package setup
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/microerror"
+)
+
+func ensureBastionHostCreated(ctx context.Context, clusterID string, config Config) error {
+	var err error
+
+	var subnetID string
+	var vpcID string
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "finding public subnet and vpc")
+
+		i := &ec2.DescribeSubnetsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("tag:giantswarm.io/cluster"),
+					Values: []*string{aws.String(clusterID)},
+				},
+				{
+					Name:   aws.String("tag:aws:cloudformation:logical-id"),
+					Values: []*string{aws.String("PublicSubnet")},
+				},
+			},
+		}
+
+		o, err := config.AWSClient.EC2.DescribeSubnets(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(o.Subnets) != 1 {
+			return microerror.Maskf(executionFailedError, "expected one subnet, got %d", len(o.Subnets))
+		}
+
+		subnetID = *o.Subnets[0].SubnetId
+		vpcID = *o.Subnets[0].VpcId
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found public subnet %#q and vpc %#q", subnetID, vpcID))
+	}
+
+	var workerSecurityGroupID string
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "finding worker security group")
+
+		i := &ec2.DescribeSecurityGroupsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("tag:giantswarm.io/cluster"),
+					Values: []*string{aws.String(clusterID)},
+				},
+				{
+					Name:   aws.String("tag:aws:cloudformation:logical-id"),
+					Values: []*string{aws.String("WorkerSecurityGroup")},
+				},
+			},
+		}
+
+		o, err := config.AWSClient.EC2.DescribeSecurityGroups(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(o.SecurityGroups) != 1 {
+			return microerror.Maskf(executionFailedError, "expected one security group, got %d", len(o.SecurityGroups))
+		}
+
+		workerSecurityGroupID = *o.SecurityGroups[0].GroupId
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found worker security group %#q", workerSecurityGroupID))
+	}
+
+	// We need to create a separate security group in order to allow SSH access to
+	// the bastion instance. The AWS API does not allow tagging the security group
+	// when creating it. That is why we need to separately create tags below, so
+	// we are able to find it later on when we want to clean it up.
+	var bastionSecurityGroupID string
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "creating bastion security group")
```

Comment:
When this function is called `ensureBastionHostCreated` and is called with a backoff logic, would it be good to check if the security group already exists, before creating it or alternatively catch for "already exists" error? Same thing for the EC2 instance. If the instance is coming up but is occasionally slow (as we evidenced during our pairing when testing this), it could then create multiple instances.

Reply:
I agree with you but I do not want to make this perfect now. What you describe did not happen so far and I am pretty reluctant to invest more time in all of this. Debugging the tenant clusters is what I still want to do and I am also thinking to not enable the bastion magic for all e2e tests by default. I would then comment the executing code and when somebody needs the functionality they can enable, use and disable it again. Then we can also make sure we do not leave crap behind and iterate later when the code really needs more fixing.